### PR TITLE
Dpp 1680 stopped versioned commitments

### PR DESCRIPTION
**Pull request by carlbm**
When there are versions of a commitment and the commitment is stopped then it was producing DLOCK_08 errors (multiple commitments). 
This fix does more to put commitments and versions of commitments into the correct time order

### CODE REVIEW

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.UnitTests/ScenarioTesting/StepDefinitions.cs [hunk 1]
```diff
@@ -86,11 +85,23 @@ public void TheValidationErrorsContain(string validationError)
         }
 
         [Then(@"There are (.*) payable datalocks for price episode (.*)")]
-        public void ThenTheDatalocksContain(int expectedNumber, string priceEpisodeIdentifier)
+        public void ThenThePayableDatalocksContain(int expectedNumber, string priceEpisodeIdentifier)
         {
             _resultsContext.DatalockValidationResult
                 .PriceEpisodePeriodMatches
                 .Where(x => x.Payable)
+                .Where(x => x.PriceEpisodeIdentifier == priceEpisodeIdentifier)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock.UnitTests/ScenarioTesting/StepDefinitions.cs:93]: should these where's be combined int 1 where clause?

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/MatcherFactory.cs [hunk 2]
```diff
@@ -6,12 +6,12 @@ public static IMatcher CreateMatcher()
         {
             return new UlnMatchHandler(
                 new UkprnMatchHandler(
-                    new StandardMatchHandler(
-                        new FrameworkMatchHandler(
-                            new PathwayMatchHandler(
-                                new ProgrammeMatchHandler(
-                                    new PriceMatchHandler(
-                                        new StartDateMatcher(
+                    new StartDateMatcher(
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/MatcherFactory.cs:9]: Does moving the StartDateMatchHandler up, is this going to break tests? and generate alternative DLOCK errors

Comment (carlbm) [reply src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Application/DataLock/Matcher/MatcherFactory.cs:9]: This is how it should have been. If the start date is wrong and there are also other errors it breaks the commitments system

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs [hunk 3]
```diff
@@ -14,65 +14,106 @@ public class LearnerCommitments
...
+        }
 
-                                          (x.EffectiveEndDate == null ||
-                                          x.EffectiveEndDate >= date) &&
+        private void RemoveCommitmentsWithNoDays(IEnumerable<Commitment> commitmentList)
+        {
+            var commitmentsToRemove = new List<Commitment>();
+            foreach (var commitment in commitmentList)
+            {
+                if (commitment.EffectiveEndDate < commitment.EffectiveStartDate)
```
Comment (Pauljgraham) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/LearnerCommitments.cs:101]: Don't we want to remove commitment.EffectiveEndDate <= commitment.EffectiveStartDate.  I think we have agreed that it should be <= but I won't update the code. Sorry, I don't have time at the moment

File: src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/Extensions/CommitmentExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿namespace SFA.DAS.CollectionEarnings.DataLock.Domain.Extensions
+{
+    public static class CommitmentExtensions
+    {
+        public static bool WithdrawnBeforeStart(this Commitment source)
+        {
+            if (source.WithdrawnOnDate < source.EffectiveFrom)
```
Comment (scottywakefield) [inline src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Domain/Extensions/CommitmentExtensions.cs]: does time matter here? 
```suggestion
            if (source.WithdrawnOnDate.Date < source.EffectiveFrom.Date)
```

**Review by scottywakefield — CHANGES_REQUESTED** [review]
I'm not sure how big a deal that date comparison is, but could cause an issue if the date's contain time as well when written to the db.

**Pull request merged**